### ISSUE TITLE
Never becomes master vrrp_startup_delay timeout

### ISSUE THREAD

**Issue by louis-oui** [acassen/keepalived#1808]
**Describe the bug**
After patch
https://github.com/acassen/keepalived/commit/1d9221721f764d991f2eeda576d693d8bd323294 for https://github.com/acassen/keepalived/issues/1803
following interfaces does not become  master :

vrrp.1
vrrp.wan
vrrp.8

Got messages.

```
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: Configuration specifies interface vlan1 which doesn't currently exist - will use if created
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: Configuration specifies interface wan which doesn't currently exist - will use if created
```
vrrp.1, vrrp.wan and vrrp.8 respectively depends on vlan1, wan and vlan8. vrrp.8 is NOK but log does not complain on it.

Log compare betwee, vrrp-ha (OK) and vrrp.wan (NOK):
```
journalctl --since "2020-12-10 12:51:00" _PID=1520 | grep vrrp-ha
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp-ha): Success creating VMAC interface vrrp-ha
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp-ha) Entering BACKUP STATE (init)
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp-ha) start preempt delay (60.000000)
Dec 10 12:52:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp-ha) received lower priority (150) advert from 192.168.20.3 - discarding
Dec 10 12:52:31 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp-ha) received lower priority (150) advert from 192.168.20.3 - discarding
Dec 10 12:52:32 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp-ha) received lower priority (150) advert from 192.168.20.3 - discarding
Dec 10 12:52:32 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp-ha) Receive advertisement timeout
Dec 10 12:52:32 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp-ha) Entering MASTER STATE
Dec 10 12:52:32 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp-ha) setting VIPs.
Dec 10 12:52:32 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp-ha) Sending/queueing gratuitous ARPs on vrrp-ha for 192.168.20.1
Dec 10 12:52:32 ipsecgw-master-patch Keepalived_vrrp[1520]: Sending gratuitous ARP on vrrp-ha for 192.168.20.1
Dec 10 12:52:32 ipsecgw-master-patch Keepalived_vrrp[1520]: Sending gratuitous ARP on vrrp-ha for 192.168.20.1
Dec 10 12:52:32 ipsecgw-master-patch Keepalived_vrrp[1520]: Sending gratuitous ARP on vrrp-ha for 192.168.20.1
Dec 10 12:52:32 ipsecgw-master-patch Keepalived_vrrp[1520]: Sending gratuitous ARP on vrrp-ha for 192.168.20.1
Dec 10 12:52:32 ipsecgw-master-patch Keepalived_vrrp[1520]: Sending gratuitous ARP on vrrp-ha for 192.168.20.1
Dec 10 12:52:37 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp-ha) Sending/queueing gratuitous ARPs on vrrp-ha for 192.168.20.1
Dec 10 12:52:37 ipsecgw-master-patch Keepalived_vrrp[1520]: Sending gratuitous ARP on vrrp-ha for 192.168.20.1
Dec 10 12:52:37 ipsecgw-master-patch Keepalived_vrrp[1520]: Sending gratuitous ARP on vrrp-ha for 192.168.20.1
Dec 10 12:52:37 ipsecgw-master-patch Keepalived_vrrp[1520]: Sending gratuitous ARP on vrrp-ha for 192.168.20.1
Dec 10 12:52:37 ipsecgw-master-patch Keepalived_vrrp[1520]: Sending gratuitous ARP on vrrp-ha for 192.168.20.1
Dec 10 12:52:37 ipsecgw-master-patch Keepalived_vrrp[1520]: Sending gratuitous ARP on vrrp-ha for 192.168.20.1
Dec 10 13:32:07 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp-ha) sent 0 priority
Dec 10 13:32:07 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp-ha) removing VIPs.
Dec 10 13:32:08 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp-ha) Success removing VMAC interface vrrp-ha
```

```
 journalctl --since "2020-12-10 12:51:00" _PID=1520 | grep vrrp.wan
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp.wan) interface wan is down
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp.wan): entering FAULT state (interface wan down)
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp.wan): entering FAULT state (interface wan down)
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp.wan): entering FAULT state (interface vrrp.wan down)
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp.wan) entering FAULT state (no IPv4 address for interface)
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp.wan) entering FAULT state
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp.wan): Success creating VMAC interface vrrp.wan
Dec 10 12:51:32 ipsecgw-master-patch Keepalived_vrrp[1520]: Netlink reports vrrp.wan up
Dec 10 13:32:08 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp.wan) Success removing VMAC interface vrrp.wan
```

Same info - see https://github.com/acassen/keepalived/issues/1803

**System Log entries**
```
Dec 10 12:51:30 ipsecgw-master-patch Keepalived[1515]: Starting Keepalived v2.1.5 (12/09,2020), git commit v2.1.5-245-g65eb6c0e
Dec 10 12:51:30 ipsecgw-master-patch Keepalived[1515]: Running on Linux 5.3.0-46-generic #38+6wind2~18.04.1 SMP Wed May 13 11:59:56 CEST 2020 (built for Linux 4.15.18)
Dec 10 12:51:30 ipsecgw-master-patch Keepalived[1515]: Command line: '/usr/sbin/keepalived' '-D'
Dec 10 12:51:30 ipsecgw-master-patch Keepalived[1515]: Opening file '/etc/keepalived/keepalived.conf'.
Dec 10 12:51:30 ipsecgw-master-patch Keepalived[1515]: Configuration file /etc/keepalived/keepalived.conf
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: Registering Kernel netlink reflector
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: Registering Kernel netlink command channel
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: "vrrp_track_file" is deprecated, please use "track_file"
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: Configuration specifies interface vlan1 which doesn't currently exist - will use if created
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: Configuration specifies interface wan which doesn't currently exist - will use if created
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: Delaying startup for 10 seconds
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp.8): Success creating VMAC interface vrrp.8
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (switch-lag) MAC Address changed from 0c:23:ac:96:67:01 to 7a:8b:51:3c:52:7c
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (switch-lag) MAC Address changed from 7a:8b:51:3c:52:7c to 0c:23:ac:96:67:01
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: Interface wan added
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp.wan) interface wan is down
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: Assigned address 192.168.20.2 for interface ha
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: Assigned address fe80::e23:acff:fe96:6703 for interface ha
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp-ha): Success creating VMAC interface vrrp-ha
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp.8): entering FAULT state (interface vlan8 down)
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp.1): entering FAULT state (interface vlan1 down)
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp.wan): entering FAULT state (interface wan down)
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp.wan): entering FAULT state (interface wan down)
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp.1): entering FAULT state (interface vrrp.1 down)
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp.wan): entering FAULT state (interface vrrp.wan down)
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp.8): entering FAULT state (interface vrrp.8 down)
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp.1) entering FAULT state (no IPv4 address for interface)
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: VRRP_Group(mygroup): Syncing vrrp.1 to FAULT state
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp.wan) entering FAULT state (no IPv4 address for interface)
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp.8) entering FAULT state (no IPv4 address for interface)
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp.1) entering FAULT state
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp.wan) entering FAULT state
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp.8) entering FAULT state
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: Registering gratuitous ARP shared channel
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp-ha) Entering BACKUP STATE (init)
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: VRRP sockpool: [ifindex(  0), family(IPv4), proto(112), fd(-1,-1)]
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: VRRP sockpool: [ifindex(  0), family(IPv4), proto(112), fd(-1,-1)]
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: VRRP sockpool: [ifindex( 14), family(IPv4), proto(112), fd(13,14)]
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: VRRP sockpool: [ifindex( 15), family(IPv4), proto(112), fd(15,16)]
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp.wan): Success creating VMAC interface vrrp.wan
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp-ha) start preempt delay (60.000000)
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (toswitch2) MAC Address changed from 0c:23:ac:96:67:02 to 0c:23:ac:96:67:01
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: Interface vlan1 added
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp.1) interface vlan1 is down
Dec 10 12:51:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp.1): Success creating VMAC interface vrrp.1
Dec 10 12:51:31 ipsecgw-master-patch Keepalived_vrrp[1520]: Assigned address 10.10.1.252 for interface vlan1
Dec 10 12:51:31 ipsecgw-master-patch Keepalived_vrrp[1520]: Assigned address 10.10.8.252 for interface vlan8
Dec 10 12:51:31 ipsecgw-master-patch Keepalived_vrrp[1520]: Assigned address 20.20.20.252 for interface wan
Dec 10 12:51:32 ipsecgw-master-patch Keepalived_vrrp[1520]: Interface dsvti1 added
Dec 10 12:51:32 ipsecgw-master-patch Keepalived_vrrp[1520]: Interface dsvti2 added
Dec 10 12:51:32 ipsecgw-master-patch Keepalived_vrrp[1520]: Netlink reports wan up
Dec 10 12:51:32 ipsecgw-master-patch Keepalived_vrrp[1520]: Netlink reports vrrp.wan up
Dec 10 12:51:32 ipsecgw-master-patch Keepalived_vrrp[1520]: Netlink reports vlan1 up
Dec 10 12:51:32 ipsecgw-master-patch Keepalived_vrrp[1520]: Netlink reports vrrp.1 up
Dec 10 12:51:32 ipsecgw-master-patch Keepalived_vrrp[1520]: Netlink reports vlan8 up
Dec 10 12:51:32 ipsecgw-master-patch Keepalived_vrrp[1520]: Netlink reports vrrp.8 up
Dec 10 12:51:32 ipsecgw-master-patch Keepalived_vrrp[1520]: Interface dsvti1 deleted
Dec 10 12:51:33 ipsecgw-master-patch Keepalived_vrrp[1520]: Assigned address fe80::788b:51ff:fe3c:527c for interface wan
Dec 10 12:51:34 ipsecgw-master-patch Keepalived_vrrp[1520]: Assigned address fe80::788b:51ff:fe3c:527c for interface vlan8
Dec 10 12:51:34 ipsecgw-master-patch Keepalived_vrrp[1520]: Assigned address fe80::788b:51ff:fe3c:527c for interface vlan1
Dec 10 12:52:30 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp-ha) received lower priority (150) advert from 192.168.20.3 - discarding
Dec 10 12:52:31 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp-ha) received lower priority (150) advert from 192.168.20.3 - discarding
Dec 10 12:52:32 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp-ha) received lower priority (150) advert from 192.168.20.3 - discarding
Dec 10 12:52:32 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp-ha) Receive advertisement timeout
Dec 10 12:52:32 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp-ha) Entering MASTER STATE
Dec 10 12:52:32 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp-ha) setting VIPs.
Dec 10 12:52:32 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp-ha) Sending/queueing gratuitous ARPs on vrrp-ha for 192.168.20.1
Dec 10 12:52:32 ipsecgw-master-patch Keepalived_vrrp[1520]: Sending gratuitous ARP on vrrp-ha for 192.168.20.1
Dec 10 12:52:32 ipsecgw-master-patch Keepalived_vrrp[1520]: Sending gratuitous ARP on vrrp-ha for 192.168.20.1
Dec 10 12:52:32 ipsecgw-master-patch Keepalived_vrrp[1520]: Sending gratuitous ARP on vrrp-ha for 192.168.20.1
Dec 10 12:52:32 ipsecgw-master-patch Keepalived_vrrp[1520]: Sending gratuitous ARP on vrrp-ha for 192.168.20.1
Dec 10 12:52:32 ipsecgw-master-patch Keepalived_vrrp[1520]: Sending gratuitous ARP on vrrp-ha for 192.168.20.1
Dec 10 12:52:37 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp-ha) Sending/queueing gratuitous ARPs on vrrp-ha for 192.168.20.1
Dec 10 12:52:37 ipsecgw-master-patch Keepalived_vrrp[1520]: Sending gratuitous ARP on vrrp-ha for 192.168.20.1
Dec 10 12:52:37 ipsecgw-master-patch Keepalived_vrrp[1520]: Sending gratuitous ARP on vrrp-ha for 192.168.20.1
Dec 10 12:52:37 ipsecgw-master-patch Keepalived_vrrp[1520]: Sending gratuitous ARP on vrrp-ha for 192.168.20.1
Dec 10 12:52:37 ipsecgw-master-patch Keepalived_vrrp[1520]: Sending gratuitous ARP on vrrp-ha for 192.168.20.1
Dec 10 12:52:37 ipsecgw-master-patch Keepalived_vrrp[1520]: Sending gratuitous ARP on vrrp-ha for 192.168.20.1
Dec 10 13:32:07 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp-ha) sent 0 priority
Dec 10 13:32:07 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp-ha) removing VIPs.
Dec 10 13:32:08 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp.1) Success removing VMAC interface vrrp.1
Dec 10 13:32:08 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp.wan) Success removing VMAC interface vrrp.wan
Dec 10 13:32:08 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp.8) Success removing VMAC interface vrrp.8
Dec 10 13:32:08 ipsecgw-master-patch Keepalived_vrrp[1520]: (vrrp-ha) Success removing VMAC interface vrrp-ha
Dec 10 13:32:08 ipsecgw-master-patch Keepalived_vrrp[1520]: Stopped - used 0.272400 user time, 0.833858 system time


```

**Comment by pqarmitage** [comment]
@louis-oui Do you mean after commit 1d92217 or after commit cf657c2? Commit 1d92217 has now been reverted, and replaced with commit cf657c2. Any testing you are doing now should include commit cf657c2.

**Comment by louis-oui** [comment]
I have tested with latest master that includes commit https://github.com/acassen/keepalived/commit/cf657c247a51b5c359e38eb746d9d6618222ccf3

`Starting Keepalived v2.1.5 (12/09,2020), git commit v2.1.5-245-g65eb6c0e
`

**Comment by louis-oui** [comment]
I added a log to debug:
```
diff --git a/keepalived/vrrp/vrrp_scheduler.c b/keepalived/vrrp/vrrp_scheduler.c
index a02186be..4d962e7c 100644
--- a/keepalived/vrrp/vrrp_scheduler.c
+++ b/keepalived/vrrp/vrrp_scheduler.c
@@ -294,7 +294,7 @@ void
 vrrp_init_instance_sands(vrrp_t *vrrp)
 {
        set_time_now();
-
+       log_message(LOG_INFO, "VRRP State of %s at vrrp_init_instance_sands %d", vrrp->iname, vrrp->state);
        if (vrrp->state == VRRP_STATE_MAST) {
                if (vrrp->reload_master)
                        vrrp->sands = time_now;

```

Got log. It never enter vrrp_init_instance_sands for NOK interfaces:
```
Dec 10 16:48:13 ipsecgw-master-patch Keepalived_vrrp[1499]: Registering Kernel netlink reflector
Dec 10 16:48:13 ipsecgw-master-patch Keepalived_vrrp[1499]: Registering Kernel netlink command channel
Dec 10 16:48:13 ipsecgw-master-patch Keepalived_vrrp[1499]: "vrrp_track_file" is deprecated, please use "track_file"
Dec 10 16:48:13 ipsecgw-master-patch Keepalived_vrrp[1499]: Configuration specifies interface vlan1 which doesn't currently exist - will use if created
Dec 10 16:48:13 ipsecgw-master-patch Keepalived_vrrp[1499]: Configuration specifies interface wan which doesn't currently exist - will use if created
Dec 10 16:48:13 ipsecgw-master-patch Keepalived_vrrp[1499]: Configuration specifies interface vlan8 which doesn't currently exist - will use if created
Dec 10 16:48:13 ipsecgw-master-patch Keepalived_vrrp[1499]: Delaying startup for 10 seconds
Dec 10 16:48:13 ipsecgw-master-patch Keepalived_vrrp[1499]: use_vmac or no_accept/strict specified, but no firewall configured - using nftables
Dec 10 16:48:13 ipsecgw-master-patch Keepalived_vrrp[1499]: Assigned address 192.168.20.2 for interface ha
Dec 10 16:48:13 ipsecgw-master-patch Keepalived_vrrp[1499]: Assigned address fe80::e23:acff:fe96:6703 for interface ha
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: (vrrp-ha): Success creating VMAC interface vrrp-ha
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: Interface wan added
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: (vrrp.wan) interface wan is down
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: (switch-lag) MAC Address changed from a6:53:ba:71:80:e3 to 0c:23:ac:96:67:01
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: (toswitch2) MAC Address changed from 0c:23:ac:96:67:02 to 0c:23:ac:96:67:01
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: (vrrp.1): entering FAULT state (interface vlan1 down)
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: (vrrp.wan): entering FAULT state (interface wan down)
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: (vrrp.wan): entering FAULT state (interface wan down)
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: (vrrp.8): entering FAULT state (interface vlan8 down)
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: (vrrp.1): entering FAULT state (interface vrrp.1 down)
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: (vrrp.wan): entering FAULT state (interface vrrp.wan down)
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: (vrrp.8): entering FAULT state (interface vrrp.8 down)
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: (vrrp.1) entering FAULT state (no IPv4 address for interface)
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP_Group(mygroup): Syncing vrrp.1 to FAULT state
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: (vrrp.wan) entering FAULT state (no IPv4 address for interface)
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: (vrrp.8) entering FAULT state (no IPv4 address for interface)
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: (vrrp.1) entering FAULT state
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: (vrrp.wan) entering FAULT state
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: (vrrp.8) entering FAULT state
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: Registering gratuitous ARP shared channel
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: (vrrp-ha) Entering BACKUP STATE (init)
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp.1 at vrrp_init_instance_sands 3
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp.wan at vrrp_init_instance_sands 3
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp.8 at vrrp_init_instance_sands 3
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP sockpool: [ifindex(  0), family(IPv4), proto(112), fd(-1,-1)]
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP sockpool: [ifindex(  0), family(IPv4), proto(112), fd(-1,-1)]
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP sockpool: [ifindex(  0), family(IPv4), proto(112), fd(-1,-1)]
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP sockpool: [ifindex( 13), family(IPv4), proto(112), fd(13,14)]
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: (vrrp.wan): Success creating VMAC interface vrrp.wan
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp.wan at vrrp_init_instance_sands 3
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: Interface vlan1 added
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: (vrrp.1) interface vlan1 is down
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: (vrrp.1): Success creating VMAC interface vrrp.1
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp.1 at vrrp_init_instance_sands 3
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: Interface vlan8 added
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: (vrrp.8) interface vlan8 is down
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: (vrrp.8): Success creating VMAC interface vrrp.8
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp.8 at vrrp_init_instance_sands 3
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: (vrrp-ha) start preempt delay (60.000000)
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:15 ipsecgw-master-patch Keepalived_vrrp[1499]: Assigned address 10.10.8.252 for interface vlan8
Dec 10 16:48:15 ipsecgw-master-patch Keepalived_vrrp[1499]: Assigned address 20.20.20.252 for interface wan
Dec 10 16:48:15 ipsecgw-master-patch Keepalived_vrrp[1499]: Assigned address 10.10.1.252 for interface vlan1
Dec 10 16:48:15 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:16 ipsecgw-master-patch Keepalived_vrrp[1499]: Netlink reports vlan8 up
Dec 10 16:48:16 ipsecgw-master-patch Keepalived_vrrp[1499]: Netlink reports vrrp.8 up
Dec 10 16:48:16 ipsecgw-master-patch Keepalived_vrrp[1499]: Netlink reports vlan1 up
Dec 10 16:48:16 ipsecgw-master-patch Keepalived_vrrp[1499]: Netlink reports vrrp.1 up
Dec 10 16:48:16 ipsecgw-master-patch Keepalived_vrrp[1499]: Netlink reports wan up
Dec 10 16:48:16 ipsecgw-master-patch Keepalived_vrrp[1499]: Netlink reports vrrp.wan up
Dec 10 16:48:16 ipsecgw-master-patch Keepalived_vrrp[1499]: Interface dsvti1 added
Dec 10 16:48:16 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:16 ipsecgw-master-patch Keepalived_vrrp[1499]: Interface dsvti1 deleted
Dec 10 16:48:16 ipsecgw-master-patch Keepalived_vrrp[1499]: Interface dsvti2 added
Dec 10 16:48:17 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:17 ipsecgw-master-patch Keepalived_vrrp[1499]: Assigned address fe80::a453:baff:fe71:80e3 for interface wan
Dec 10 16:48:18 ipsecgw-master-patch Keepalived_vrrp[1499]: Assigned address fe80::a453:baff:fe71:80e3 for interface vlan8
Dec 10 16:48:18 ipsecgw-master-patch Keepalived_vrrp[1499]: Assigned address fe80::a453:baff:fe71:80e3 for interface vlan1
Dec 10 16:48:18 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:19 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:20 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:21 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:22 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:23 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:24 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:25 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:26 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:27 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:28 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:29 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:30 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:31 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:32 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:33 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:34 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:35 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:36 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:37 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:38 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:39 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:40 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:41 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:42 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:43 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:44 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:45 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:46 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:47 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:48 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:49 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:50 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1
Dec 10 16:48:51 ipsecgw-master-patch Keepalived_vrrp[1499]: VRRP State of vrrp-ha at vrrp_init_instance_sands 1

```

**Comment by pqarmitage** [comment]
I have tried reproducing what appears to be happening, and I am not experiencing the problem that you are. In order to collect more detailed information so that I can work out what is happen, I have set out below how to collect the further information I need.

Can you please run the following script before you start keepalived and leave it running while keepalived is running:
```
#!/bin/bash

ip link show >/tmp/ip.log
ip addr show >>/tmp/ip.log
ip monitor all >>/tmp/ip.log
```

then start keepalived (if possible with the `-d` command line option added), and allow the interfaces to come up etc. Once everything is settled, but while vrrp.1, vrrp.8 and vrrp.wan are still down, execute
`kill -USR1 $(cat /run/keepalived.pid)`
which will create a file /tmp/keepalived.data

At this point, kill the script still running ip monitor all, and then do
```
ip link show >>/tmp/ip.log
ip addr show >>/tmp/ip.log
```

and finally either post here or email to me directly the /tmp/keepalived.data and /tmp/ip.log files and all the syslog output for keepalived from when it starts up to when it terminates, and I should then be able to see what is happening.

**Comment by pqarmitage** [comment]
Commit deafd99 resolves this issue.

The problem was indicated by the following log entries:
```
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: (vrrp-ha): Success creating VMAC interface vrrp-ha
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: Interface wan added
Dec 10 16:48:14 ipsecgw-master-patch Keepalived_vrrp[1499]: (vrrp.wan) interface wan is down
```
keepalived increments num_script_if_fault when it receives notification via netlink that a link (network interface) of a vrrp instance is down, but the initialisation of num_script_if_fault isn't done until after the VMAC interfaces are created. So if keepalived receives notification of a new interface (which will be down) while it is creating the VMACs, it increments num_script_if_fault before the variable is set, and so num_script_if_fault ends up having a value too large by 1, and so when the interface eventually is set up, num_script_if_fault ends up with the value 1, and it is still considered to be in the fault state.

On my development system, I couldn't get the interface to be created quickly enough so that it occurred while the VMAC interfaces were being create, so I had to add a delay in the keepalived code around VMAC creation in order to be able to reproduce the problem. I think probably for the problem to occur a system with several processor cores is needed and I don't have an x86_64 system with more than 2 cores.

**Comment by louis-oui** [comment]
Thank you @pqarmitage. It solves the issue. I did not notice the log interface wan down after being added.

My tests was a two cores machine. Anyway, that's a good news the problem is sorted out.
